### ISSUE TITLE
Incorrect "Redundant brackets" because of missing fixity declaration

### ISSUE THREAD

**Issue by reinerp** [ndmitchell/hlint#426]
Consider these two input files:

```haskell
module HlintRedundantBracketsBase where

infixl 6 +?

(+?) :: a -> a -> a
(+?) = undefined
```

and

```haskell
module HlintRedundantBrackets where

import HlintRedundantBracketsBase((+?))

bad x = x . (x +? x . x)


infixl 6 +??

(+??) :: a -> a -> a
(+??) = undefined

good x = x . (x +?? x . x)
```

HLint warns about the definition of `bad`, but not the definition of `good`:

```
HlintRedundantBrackets.hs:5:9: Suggestion: Redundant bracket
Found:
  x . (x +? x . x)
Why not:
  x . x +? x . x
```

It seems like HLint is assuming the incorrect fixity of `+?` because it doesn't read the imported file to find the true fixity.

A (rather brutal) fix would be to suppress this error for operators that aren't in Prelude and don't have fixity definitions in the current file.

**Comment by ndmitchell** [comment]
Thanks for the report. The bit that goes "wrong" is `needBracket 1 (x . (x +? x . x)) (x +? x . x)` on at https://github.com/ndmitchell/hlint/blob/fbfc44e334cce86a887dfa4301c7a6edd4c74df9/src/Hint/Bracket.hs#L117. That basically asks if `x . ???` needs a bracket when `???` is `x +? x . x`. If you have internally bracketed the `x +? x . x` as `(x +? x) . x` then the answer is no, you don't need a bracket, which is what haskell-src-exts-util answers: https://github.com/pepeiborra/haskell-src-exts-util/blob/a8b3395202d7d0dc3fec2d682633e06367cae554/src/Language/Haskell/Exts/Bracket.hs#L73

The correct thing to do is for HLint to declare that "hmm, if I had misparsed this, then the bracket would be required, so don't be too eager to avoid it". I can add that as a special case in removing bracket code. However, it occurs to me that the rest of haskell-src-exts-util bracket code isn't making any assumptions about relative fixities of dot/dollar or anything else, so having that one rule looks a bit odd (and is also incorrect if people have redefined `.`). Perhaps it should be removed entirely from `exts-util` and the custom/hacky nature of when/when-not to bracket should be put into HLint itself as domain-specific tweaks. Thoughts @pepeiborra

**Comment by pepeiborra** [comment]
Agreed, that rule should be removed from `haskell-src-exts-util` and moved back into `hlint`. I was about to make a release of `haskell-src-exts-util` for GHC 8.4 compatibility, so I will take the chance to remove it. (but let me know if you prefer me to do this later)

The original motivation to extract haskell-src-exts-util from hlint was so that other projects like [arrowp-qq](http://github.com/pepeiborra/arrowp) could reuse it. That hasn't changed a bit, the code for rebracketing is still generally useful. And it follows that it should not make any internal assumptions, as you said. It would be fine if the assumptions were provided externally as e.g. a set of fixity declarations.

**Comment by ndmitchell** [comment]
Please remove it now, that sounds good.

I think rebracketing with respect to fixities is perhaps something that code live in HSEU, but it's not something arrow-qq or hlint requires, so something to wait for a need.

**Comment by ndmitchell** [comment]
With haskell-src-exts-util-0.2.2 the bug goes away - it now never tries to remove brackets based on fixities. I did need to copy some of the old code over for some other place where I was inserting brackets unnecessarily, so there's a bit more duplication than ideal, but I'll sort that out in a future refactor.

**Comment by ndmitchell** [comment]
@reinerp to upgrade, make sure you upgrade haskell-src-exts-util first, then upgrade hlint (once I release).
